### PR TITLE
Apollo as a ConfigCenter.

### PR DESCRIPTION
**Pull request by dongzl**
Apollo as a ConfigCenter.

### CODE REVIEW

File: sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstance.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    private Properties properties = new Properties();
+    
+    @Override
+    public void init(final InstanceConfiguration config) {
+        initApolloConfig(config);
+        initApolloOpenApiClient();
+        initKeysRelationship();
+    }
+    
+    private void initApolloConfig(final InstanceConfiguration config) {
```
Comment (wgy8283335) [inline sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstance.java]: Shall we rename the method name, as the method doesn't only initialize the "apollyConfig".

File: sharding-orchestration/sharding-orchestration-center/src/test/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstanceTest.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public static void init() {
+        Config apolloConfig = mock(Config.class);
+        when(apolloConfig.getProperty("test.children.1", "")).thenReturn("value1");
+        FieldSetter.setField(configCenter, ApolloInstance.class.getDeclaredField("apolloConfig"), apolloConfig);
+        Set<String> instanceKeys = Sets.newHashSet();
+        instanceKeys.add("test.children.1");
+        instanceKeys.add("test.children.2");
+        instanceKeys.add("test1.children.3");
+        ConfigTreeNode root = new ConfigTreeNode(null, "/", Sets.<ConfigTreeNode>newHashSet());
+        root.initTree(instanceKeys, ".");
```
Comment (wgy8283335) [inline sharding-orchestration/sharding-orchestration-center/src/test/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstanceTest.java]: Shall we provide one initialize method to make creating a config tree node more easily, as we need two lines now?

File: sharding-orchestration/sharding-orchestration-center/src/test/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstanceTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.Set;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public final class ApolloInstanceTest {
```
Comment (wgy8283335) [inline sharding-orchestration/sharding-orchestration-center/src/test/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstanceTest.java]: Do we need unit tests for watch, close, getType?

File: sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+@Getter
+@Setter
+@AllArgsConstructor
+public final class ConfigTreeNode {
+    
+    private static final String SHARDING_SPHERE_KEY_ROOT = "/";
+    
+    private static final String SHARDING_SPHERE_KEY_SEPARATOR = "/";
+    
+    private static Map<String, Set<String>> childrenKeyMap = new ConcurrentHashMap<>();
```
Comment (wgy8283335) [inline sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java]: Is this Map only store the children of this node? If it's true, why not use Set<String> directly?

File: sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    
+    private Set<ConfigTreeNode> childrenNodes;
+    
+    /**
+     * init tree.
+     * 
+     * @param instanceKeys instance Key set
+     * @param keySeparator key separator
+     */
+    public void initTree(final Set<String> instanceKeys, final String keySeparator) {
```
[the remark this answers is not in the document]
Comment (dongzl) [reply sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java]: OK, I fix it.

File: sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            return;
+        }
+        for (String each : childrenKeys) {
+            ConfigTreeNode child = new ConfigTreeNode(parentNode, each, Sets.<ConfigTreeNode>newHashSet());
+            parentNode.getChildrenNodes().add(child);
+            initTree(child, each);
+        }
+    }
+    
+    private void initKeysRelationship(final Set<String> instanceKeys, final String keySeparator) {
```
[the remark this answers is not in the document]
Comment (dongzl) [reply sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java]: OK.

File: sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstance.java (hunk 1 above)
Comment (dongzl) [reply sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstance.java]: This is only a `private` method and inner used. The `public` method is `init(final InstanceConfiguration config)`.

File: sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        for (String each : instanceKeys) {
+            initKeysRelationship(each, keySeparator);
+        }
+    }
+    
+    private void initKeysRelationship(final String instanceKey, final String keySeparator) {
+        if (!instanceKey.contains(keySeparator)) {
```
[the remark this answers is not in the document]
Comment (dongzl) [reply sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java]: Yes, I add this sentence's unit test at `ApolloInstanceTest` class.

File: sharding-orchestration/sharding-orchestration-center/src/test/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstanceTest.java (hunk 3 above)
Comment (dongzl) [reply sharding-orchestration/sharding-orchestration-center/src/test/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstanceTest.java]: I'm sorry, Those methods are unit test unfriendly, `watch` is Apollo‘s `ConfigChangeListener`, now I make unit test as Mockito's mock method, It's a little difficult. I want to test it in dev mode.

`close` only a blank method, There’s no need `close` method for Apollo.

File: sharding-orchestration/sharding-orchestration-center/src/test/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstanceTest.java (hunk 2 above)
Comment (dongzl) [reply sharding-orchestration/sharding-orchestration-center/src/test/java/org/apache/shardingsphere/orchestration/center/instance/ApolloInstanceTest.java]: It's a good suggestion? But what's your concrete suggestion? How to make it more easily?

used static method?

File: sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java (hunk 4 above)
Comment (dongzl) [reply sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java]: No, The `Map` store the relationships between all the key and it's children's keys, so I use the Map structure.

File: sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java [hunk 8]
```diff
@@ -53,29 +53,29 @@
...
-    public void initTree(final Set<String> instanceKeys, final String keySeparator) {
-        initKeysRelationship(instanceKeys, keySeparator);
-        initTree(this, SHARDING_SPHERE_KEY_ROOT);
+    public void init(final Set<String> instanceKeys, final String keySeparator) {
+        initKeysRelationships(instanceKeys, keySeparator);
+        init(this, SHARDING_SPHERE_KEY_ROOT);
     }
     
-    private void initTree(final ConfigTreeNode parentNode, final String shardingSphereKey) {
+    private void init(final ConfigTreeNode parentNode, final String shardingSphereKey) {
```
Comment (wgy8283335) [inline sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java]: This should use another name, different from init().

Comment (dongzl) [reply sharding-orchestration/sharding-orchestration-center/src/main/java/org/apache/shardingsphere/orchestration/center/instance/node/ConfigTreeNode.java]: fix it, initTreeNode()

**Pull request merged**